sql/core/src/main/scala/org/apache/spark/sql/functions.scala

```diff
@@ -3055,13 +3056,21 @@ object functions {
    * with the specified schema. Returns `null`, in the case of an unparseable string.
    *
    * @param e a string column containing JSON data.
-   * @param schema the schema to use when parsing the json string as a json string
+   * @param schema the schema to use when parsing the json string as a json string. In Spark 2.1,
+   *               the user-provided schema has to be in JSON format. Since Spark 2.2, the DDL
+   *               format is also supported for the schema.
    *
    * @group collection_funcs
    * @since 2.1.0
    */
-  def from_json(e: Column, schema: String, options: java.util.Map[String, String]): Column =
-    from_json(e, DataType.fromJson(schema), options)
+  def from_json(e: Column, schema: String, options: java.util.Map[String, String]): Column = {
+    val dataType = try {
+      DataType.fromJson(schema)
+    } catch {
+      case NonFatal(_) => StructType.fromDDL(schema)
+    }
```

Comment:
I am wondering why parsing schema from a string is implemented here but not inside of the `JsonToStructs` expression? So, calling of `from_json` in SQL and in Scala/API has different behaviour, right. Did you do that intentionally?

Reply:
I think you came to a wrong PR. The schema parsing was started in Scala side at https://github.com/maropu/spark/commit/fe33121a53384811a8e094ab6c05dc85b7c7ca87#diff-b5e6d03d9c9afbfa925e039c48e31078608ea749c193e6af3087b79eb701bc7cR2877. I guess reason is that, it was not intended to be used an expression before. Now we take it in SQL as well.

Reply:
I think I just missed it, so making the code in common seems fine.

Reply:
Thanks, @HyukjinKwon. Ah, I see, I totally forgot it...

Reply:
I moved parsing to a common place. Please, take a look at https://github.com/apache/spark/pull/30201